File: examples/lua/x18.lua
--[[
	3-d line and point plot demo.  Adapted from x08c.c.

  Copyright (C) 2008  Werner Smekal

  This file is part of PLplot.

  PLplot is free software you can redistribute it and/or modify
  it under the terms of the GNU Library General Public License as published
  by the Free Software Foundation either version 2 of the License, or
  (at your option) any later version.

  PLplot is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU Library General Public License for more details.

  You should have received a copy of the GNU Library General Public License
  along with PLplot if not, write to the Free Software
  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA
--]]

-- initialise Lua bindings for PLplot examples.
dofile("plplot_examples.lua")

function test_poly(k)
  draw= { { 1, 1, 1, 1 },
          { 1, 0, 1, 0 },
          { 0, 1, 0, 1 },
          { 1, 1, 0, 0 } }

  x = {}
  y = {}
  z = {}

  pl.adv(0)
  pl.vpor(0, 1, 0, 0.9)
  pl.wind(-1, 1, -0.9, 1.1)
  pl.col0(1)
  pl.w3d(1, 1, 1, -1, 1, -1, 1, -1, 1, alt[k], az[k])
  pl.box3("bnstu", "x axis", 0, 0,
          "bnstu", "y axis", 0, 0,
          "bcdmnstuv", "z axis", 0, 0)

  pl.col0(2)

  -- x = r sin(phi) cos(theta)
  -- y = r sin(phi) sin(theta)
  -- z = r cos(phi)
  -- r = 1 :=)

  for i=0, 19 do
    for j=0, 19 do
	    x[1] = math.sin( math.pi*j/20.1 ) * math.cos( 2*math.pi*i/20 )
	    y[1] = math.sin( math.pi*j/20.1 ) * math.sin( 2*math.pi*i/20 )
	    z[1] = math.cos( math.pi*j/20.1 )

	    x[2] = math.sin( math.pi*(j+1)/20.1 ) * math.cos( 2*math.pi*i/20 )
	    y[2] = math.sin( math.pi*(j+1)/20.1 ) * math.sin( 2*math.pi*i/20 )
	    z[2] = math.cos( math.pi*(j+1)/20.1 )

	    x[3] = math.sin( math.pi*(j+1)/20.1 ) * math.cos( 2*math.pi*(i+1)/20 )
	    y[3] = math.sin( math.pi*(j+1)/20.1 ) * math.sin( 2*math.pi*(i+1)/20 )
	    z[3] = math.cos( math.pi*(j+1)/20.1 )

	    x[4] = math.sin( math.pi*j/20.1 ) * math.cos( 2*math.pi*(i+1)/20 )
	    y[4] = math.sin( math.pi*j/20.1 ) * math.sin( 2*math.pi*(i+1)/20 )
	    z[4] = math.cos( math.pi*j/20.1 )

	    x[5] = math.sin( math.pi*j/20.1 ) * math.cos( 2*math.pi*i/20 )
	    y[5] = math.sin( math.pi*j/20.1 ) * math.sin( 2*math.pi*i/20 )
	    z[5] = math.cos( math.pi*j/20.1 )

	    pl.poly3( x, y, z, draw[k], 1 )
    end
  end

  pl.col0(3)
  pl.mtex("t", 1, 0.5, 0.5, "unit radius sphere" )
end


----------------------------------------------------------------------------
-- main
--
-- Does a series of 3-d plots for a given data set, with different
-- viewing options in each plot.
----------------------------------------------------------------------------

NPTS = 1000
opt = { 1, 0, 1, 0 }
alt = { 20, 35, 50, 65 }
az  = { 30, 40, 50, 60 }


-- Parse and process command line arguments
pl.parseopts(arg, pl.PL_PARSE_FULL)

-- Initialize plplot
pl.init()

for k=1, 4 do
	test_poly(k)
end

x = {}
y = {}
z = {}

-- From the mind of a sick and twisted physicist...
for i=1, NPTS do
  z[i] = -1 + 2*(i-1)/NPTS

  -- Pick one ...
  --	r  = 1 - (i-1) / NPTS
  r = z[i]

  x[i] = r * math.cos( 12*math.pi*(i-1)/NPTS )
  y[i] = r * math.sin( 12*math.pi*(i-1)/NPTS )
end

for k=1, 4 do
	pl.adv(0)
	pl.vpor(0, 1, 0, 0.9)
	pl.wind(-1, 1, -0.9, 1.1)
	pl.col0(1)
	pl.w3d(1, 1, 1, -1, 1, -1, 1, -1, 1, alt[k], az[k])
	pl.box3("bnstu", "x axis", 0, 0,
	        "bnstu", "y axis", 0, 0,
	        "bcdmnstuv", "z axis", 0, 0)

	pl.col0(2)

	if opt[k]~=0 then
		pl.line3( x, y, z )
	else
		-- U+22C5 DOT OPERATOR.
		pl.string3( x, y, z, "⋅" )
	end

	pl.col0(3)
	pl.mtex("t", 1.0, 0.5, 0.5, "#frPLplot Example 18 - Alt=" .. alt[k] .. ", Az=" .. az[k])
end

pl.plend()

File: examples/ocaml/x18.ml
(*
  3-d line and point plot demo.  Adapted from x08c.c.
*)

open Plplot
open Printf

let opt = [|1; 0; 1; 0|]
let alt = [|20.0; 35.0; 50.0; 65.0|]
let az = [|30.0; 40.0; 50.0; 60.0|]

let npts = 1000

let pi = atan 1.0 *. 4.0

let test_poly k =
  let draw = [| [| true;  true;  true;  true|];
                [| true; false;  true; false|];
                [|false;  true; false;  true|];
                [| true;  true; false; false|] |] in

  let two_pi = 2.0 *. pi in

  pladv 0;
  plvpor 0.0 1.0 0.0 0.9;
  plwind (-1.0) 1.0 (-0.9) 1.1;
  plcol0 1;
  plw3d 1.0 1.0 1.0 (-1.0) 1.0 (-1.0) 1.0 (-1.0) 1.0 alt.(k) az.(k);
  plbox3 "bnstu" "x axis" 0.0 0
         "bnstu" "y axis" 0.0 0
         "bcdmnstuv" "z axis" 0.0 0;

  plcol0 2;

  let theta a = two_pi *. float_of_int a /. 20.0 in
  let phi a = pi *. float_of_int a /. 20.1 in

  let x = Array.make 5 0.0 in
  let y = Array.make 5 0.0 in
  let z = Array.make 5 0.0 in

  for i = 0 to 19 do
    for j = 0 to 19 do
      x.(0) <- sin (phi j) *. cos (theta i);
      y.(0) <- sin (phi j) *. sin (theta i);
      z.(0) <- cos (phi j);

      x.(1) <- sin (phi (j + 1)) *. cos (theta i);
      y.(1) <- sin (phi (j + 1)) *. sin (theta i);
      z.(1) <- cos (phi (j + 1));

      x.(2) <- sin (phi (j + 1)) *. cos (theta (i + 1));
      y.(2) <- sin (phi (j + 1)) *. sin (theta (i + 1));
      z.(2) <- cos (phi (j + 1));

      x.(3) <- sin (phi j) *. cos (theta (i+1));
      y.(3) <- sin (phi j) *. sin (theta (i+1));
      z.(3) <- cos (phi j);

      x.(4) <- sin (phi j) *. cos (theta i);
      y.(4) <- sin (phi j) *. sin (theta i);
      z.(4) <- cos (phi j);

      plpoly3 x y z draw.(k) true;
    done
  done;

  plcol0 3;
  plmtex "t" 1.0 0.5 0.5 "unit radius sphere";
  ()

(*--------------------------------------------------------------------------*\
 * Does a series of 3-d plots for a given data set, with different
 * viewing options in each plot.
\*--------------------------------------------------------------------------*)
let () =
  (* Parse and process command line arguments *)
  plparseopts Sys.argv [PL_PARSE_FULL];

  (* Initialize plplot *)
  plinit ();

  for k = 0 to 3 do
    test_poly k
  done;

  (* From the mind of a sick and twisted physicist... *)
  let z =
    Array.init npts (
      fun i ->
        ~-.1.0 +. 2.0 *. float_of_int i /. float_of_int npts
    )
  in
  let x =
    Array.mapi (
      fun i r ->
        r *. cos (2.0 *. pi *. 6.0 *. float_of_int i /. float_of_int npts)
    ) z
  in
  let y =
    Array.mapi (
      fun i r ->
        r *. sin (2.0 *. pi *. 6.0 *. float_of_int i /. float_of_int npts)
    ) z
  in

  for k = 0 to 3 do
    pladv 0;
    plvpor 0.0 1.0 0.0 0.9;
    plwind (-1.0) 1.0 (-0.9) 1.1;
    plcol0 1;
    plw3d 1.0 1.0 1.0 (-1.0) 1.0 (-1.0) 1.0 (-1.0) 1.0 alt.(k) az.(k);
    plbox3 "bnstu" "x axis" 0.0 0
           "bnstu" "y axis" 0.0 0
           "bcdmnstuv" "z axis" 0.0 0;

    plcol0 2;

    if opt.(k) <> 0 then
      plline3 x y z
    else
      (* U+22C5 DOT OPERATOR. *)
      plstring3 x y z "⋅";

    plcol0 3;
    let title =
      sprintf "#frPLplot Example 18 - Alt=%.0f, Az=%.0f" alt.(k) az.(k)
    in
    plmtex "t" 1.0 0.5 0.5 title;
  done;

  (* Clean up *)
  plend ();
  ()

